---
title: "ftshi3-1 microbiome"
author: "Daniel Caddell"
date: "11/02/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

## load packages
```{r}
library(phyloseq); packageVersion("phyloseq") #‘1.30.0’
library(ggplot2); packageVersion("ggplot2") #‘3.2.1’
library(reshape2); packageVersion("reshape2") #‘1.4.3’
library(dplyr); packageVersion("dplyr") #‘0.8.3’
library(RColorBrewer); packageVersion("RColorBrewer") #‘1.1.2’
library(scales); packageVersion("scales") #‘1.1.0’
library(grid); packageVersion("grid") #‘3.6.1’
library(ape); packageVersion("ape") #‘5.3’
library(ggalt); packageVersion("ggalt") #‘0.4.0’
library(mctoolsr); packageVersion("mctoolsr") # ‘0.1.1.2’
library(vegan); packageVersion("vegan") # ‘2.5.6’
library(data.table); packageVersion("data.table") #‘1.12.8’
library(agricolae);packageVersion("agricolae") #‘1.3.1’
library(labdsv);packageVersion("labdsv") #
library(indicspecies);packageVersion("indicspecies") #
library(patchwork);packageVersion("patchwork") #
library(cowplot);packageVersion("cowplot") #
library(beepr); packageVersion("beepr") #‘1.3’
theme_set(theme_bw()) #Define a default theme for ggplot graphics.

#setwd("/set/working/directory/location/")
setwd("/users/dcaddell/Desktop/Laxmi/ftshi3/")
```

## import ASV biom file and sample file, then build phyloseq object  
```{r}
biom_file <- "merged-feature-table.biom"
map_file <- "metadata-ftshi3.txt"
otutax <- import_biom(biom_file, parseFunction = parse_taxonomy_default)
sam <- import_qiime_sample_data(map_file)
all <- merge_phyloseq(otutax, sam)
all
colnames(tax_table(all)) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
colnames(tax_table(all))
```

## Remove mitochondrial and chloroplast reads from a phyloseq object (Illumina - greengenes)  
A full list of related functions are in 'functions_for16SData.Rmd'  
WARNING: only use on taxonomies generated by greengenes, otherwise pattern matching will not find mito/chloro ASVs/OTUs.
This chunk has been modified from code originally written by Alex Styer.
```{r}
remove_mitoAndChloro <- function(x){#where x is a merged phyloseq object
  df <- as.data.frame(tax_table(x))
  setDT(df, keep.rownames = TRUE)
  colnames(df)[1] <- "OTU"
  allASVs <- df[,1]
  mito <- filter(df, df$Family == "f__mitochondria")
  chloro <- filter(df, df$Class == "c__Chloroplast")
  contam <- as.vector(rbind(mito, chloro)$OTU)
  ASVs_toKeep <- as.vector(subset(allASVs, !(OTU %in% contam))$OTU)
  num_mito <- nrow(mito)
  num_chloro <- nrow(chloro)
  cat(sprintf("%i and %s ASVs were removed as pesky mitochondria or chloroplast, respectively.\nKhaleesi frees these bacterial slaves to the wild binary yonder.", num_mito, num_chloro))
  x <- prune_taxa(ASVs_toKeep, x)
}
```

## Clean up taxonomy
Takes a psmelted phyloseq object and cleans up greengenes-format taxonomies.  
i.e. removes the "k__" in k__Bacteria  
Also takes NA values and unassigned taxa and renames their taxonomy as "Unknown".
This chunk has been modified from code originally written by Alex Styer.
```{r}
cleanup_taxonomy <- function(x){ #to be used on dfs made from melted phyloseq objects
  x$Domain <- as.character(lapply(x$Domain, sub, pattern = "k__", replacement = ""))
  x$Phylum <- as.character(lapply(x$Phylum, sub, pattern = "p__", replacement = ""))
  x$Class <- as.character(lapply(x$Class, sub, pattern = "c__", replacement = ""))
  x$Order <- as.character(lapply(x$Order, sub, pattern = "o__", replacement = ""))
  x$Family <- as.character(lapply(x$Family, sub, pattern = "f__", replacement = ""))
  x$Genus <- as.character(lapply(x$Genus, sub, pattern = "g__", replacement = ""))
  x$Species <- as.character(lapply(x$Species, sub, pattern = "s__", replacement = ""))
  x$Domain <- ifelse(x$Domain == "", "Unknown", as.character(x$Domain))
  x$Phylum <- ifelse(x$Phylum == "", "Unknown", as.character(x$Phylum))
  x$Class <- ifelse(x$Class == "", "Unknown", as.character(x$Class))
  x$Order <- ifelse(x$Order == "", "Unknown", as.character(x$Order))
  x$Family <- ifelse(x$Family == "", "Unknown", as.character(x$Family))
  x$Genus <- ifelse(x$Genus == "", "Unknown", as.character(x$Genus))
  x$Species <- ifelse(x$Species == "", "Unknown", as.character(x$Species))
  x$Domain <- ifelse(is.na(x$Domain) == TRUE , "Unknown", as.character(x$Domain))
  x$Phylum <- ifelse(is.na(x$Phylum) == TRUE , "Unknown", as.character(x$Phylum))
  x$Class <- ifelse(is.na(x$Class) == TRUE , "Unknown", as.character(x$Class))
  x$Order <- ifelse(is.na(x$Order) == TRUE , "Unknown", as.character(x$Order))
  x$Family <- ifelse(is.na(x$Family) == TRUE , "Unknown", as.character(x$Family))
  x$Genus <- ifelse(is.na(x$Genus) == TRUE , "Unknown", as.character(x$Genus))
  x$Species <- ifelse(is.na(x$Species) == TRUE , "Unknown", as.character(x$Species))
  x <- x
}
```

Now run the cleanup functions to remove any remaining mitochondria and chloroplast sequences.  
Also melt phyloseq object into a dataframe and transform sample counts for further analyses.  
```{r}
all.clean <- remove_mitoAndChloro(all)
all.clean.df <- psmelt (all.clean)
all.clean.df <- cleanup_taxonomy(all.clean.df)
#transform data
all.trans <- transform_sample_counts(all.clean, function(OTU) 100*OTU/sum(OTU))
#merge groups within transformed data
all.trans.merge <- merge_samples(all.trans, "Group")
all.trans.merge <- transform_sample_counts(all.trans.merge, function(OTU) 100*OTU/sum(OTU))
all.tm.df<- psmelt(all.trans.merge)
all.tm.df <- cleanup_taxonomy(all.tm.df) #NOTE: some columns will be messed up during merge_samples, which may matter for some downstream analyses.
beep(4)
```

## Save the session so that it can be loaded later without having to rerun prior steps.
```{r eval=FALSE, include=FALSE}
save.image("phyloseq_illumina.rdata")
#close R, re-open R
load("phyloseq_illumina.rdata")
```

## Alpha Diversity   
Note: Alpha diversity is being performed on rarefied data, which differs from the transformed data above (used for subsequent analyses).
```{r}
set.seed(42)
all_rare<-rarefy_even_depth(all.clean, rngseed = TRUE)
#sanity check
par(mfrow = c(1, 2))
title = "Sum of reads for each sample, all_rare"
plot(sort(sample_sums(all_rare), TRUE), type = "h", main = title, ylab = "reads") 
#, ylim = c(0, 1000))

richness = estimate_richness(all_rare,measures = "Shannon")
richness
```

ANOVA of alpha-diversity (and Tukey-HSD posthoc test)
```{r}
temp.aov <- sample_data(all_rare)
data.aov <- cbind(temp.aov, richness)
data.aov 
#ANOVA posthoc test: Tukey-HSD
all_aov <- aov(Shannon ~Group, data = data.aov)
summary(all_aov) #F-value = 0,942, alpha diversity is not significant!
#all_aov_tukey <- TukeyHSD(all_aov, 'Group', conf.level=0.95)
#all_aov_tukey 

#Plot Alpha diversity (Shannon)
plot1 <- plot_richness(all_rare, x="Group", measures=c("Shannon"), color="Group")+
        geom_point( size = 6) +
        geom_point(shape = 1,size = 6,colour = "black")+
        ylim(4.2,6.3)+
        scale_color_manual(values=c("#cb9e59","#a04344","#427e83","#4d5b6a","#ecc363","#bf7a8f","#84b59c","#7997a1"))+
        theme(axis.text.x=element_text(size=12,color="black",angle=90), 
              axis.text.y=element_text(size=12,color="black"), 
              axis.title=element_text(size=12,face="bold"), 
              text=element_text(size=12),
              legend.title=element_text(size=12), 
              legend.text=element_text(size=12))
plot1
#ggsave("alphaD.pdf", plot = plot1, width=6, height=3, useDingbats=FALSE)
#ggsave("alphaD.png", plot = plot1, width=6, height=3, dpi=600)
```

## Beta Diversity
CAP analysis
```{r}
ordcap = ordinate(all.trans, "CAP", "bray", ~Treatment*Genotype)

plot2 <- phyloseq::plot_ordination(all.trans, ordcap, "samples", color="Group", title="Bray-Curtis distance")+
          geom_vline(xintercept = 0,linetype="longdash",color="gray")+
          geom_hline(yintercept = 0,linetype="longdash",color="gray")+
    geom_point( size = 7)+
          geom_point(shape = 1,size = 7,colour = "black")+
          scale_color_manual(values=c("#cb9e59","#a04344","#427e83","#4d5b6a","#ecc363","#bf7a8f","#84b59c","#7997a1"))+
          scale_x_reverse()+
          theme(axis.text.x=element_text(size=11,color="black",angle=0), 
                axis.text.y=element_text(size=11,color="black"), 
                axis.title=element_text(size=11,face="bold"), 
                text=element_text(size=11, face="bold"),
                plot.title=element_text(size=15,hjust= 0.5, vjust=2),
                legend.title=element_text(size=10), 
                legend.text=element_text(size=8))
plot2
#ggsave("betaD.pdf", plot = plot2, width=6, height=5, useDingbats=FALSE)
#ggsave("betaD.png", plot = plot2, width=6, height=5, dpi=600)
```

Perform analysis of variance using distance matrices (adonis test and pairwise permanova posthoc test)
```{r}
ps.prop_dist <- distance(all.trans, method = "bray")
ps.prop_data <- data.frame(sample_data(all.trans))
summary(ps.prop_data)

test <- adonis(ps.prop_dist ~ Treatment*Genotype, data = ps.prop_data, permutations = 10000)
test
test2 <-  adonis(ps.prop_dist ~ Group, data = ps.prop_data, permutations = 10000)
test2
# get test results as a table
test_tab <- test$aov.tab
test_tab_rows <- rownames(test_tab)
test_tab_rows
for(i in 1:5){ #change "1:n" to number of rows being assessed
  a1 <- c("The percent of variance attributable to the factor")
  b2 <- toString(test_tab_rows[i])
  c3 <- c("is")
  d4 <- toString(round(test_tab$SumsOfSqs[i]/test_tab$SumsOfSqs[5], 5)) #change [n] to the row of "Total"
  print(paste(a1,b2,c3,d4))
}
#adonis posthoc test: Pairwise permanova
T1<-calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Treatment")
T1
G1<-calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Genotype")
G1
pm1 <- calc_pairwise_permanovas(ps.prop_dist, ps.prop_data, "Group")
pm1
write.csv(pm1,"pm1.csv")
```

## Relative abundance
For this manuscript, we have grouped ASVs by Phyla (top 5, all others merged into "Other")
```{r message=FALSE, warning=FALSE}
# Agglomerate Taxa 
glom_Phylum <- tax_glom(all.trans, taxrank = "Phylum")
glom_Phylum
## pick a taxonomic level
taxrank_level = "Phylum"
gloms <- get(paste0("glom_", taxrank_level))
dim(otu_table(gloms))[1]
gloms
## get top 5 phyla
others <- names(sort(taxa_sums(gloms), decreasing = TRUE)[6:length(names(taxa_sums(gloms)))]) # all taxa after the top 5  will be merged into "others". Change "6" if you want a different number taxa.
gloms <- merge_taxa(gloms, others, 1)
gloms
# sort based on relative abundance
# get otu table
ratab <- data.frame(otu_table(gloms))
dim(ratab)
# calculate average relative abundance for all the samples
ratab <- data.frame(rowMeans(ratab))
colnames(ratab) <- "Relative_Abundance"
# merge otu table with taxonomy table
ratab <- merge(as.data.frame(gloms@tax_table@.Data), ratab, by = "row.names")
# sorting taxa by abundance from high to low for RA
sort0 <- c()
sort0 <- ratab[with(ratab, order(ratab$Relative_Abundance)),][[taxrank_level]]
sort0 <- as.vector(sort0)
sort0[is.na(sort0)] <- "Other"
sort0 <- sort0[! sort0 %in% "Other"]
sort0 <- c("Other", sort0)
sort0
## melt dataframe
melts <- psmelt(gloms)
# convert Phyla to a character vector from a factor 
str(melts[[taxrank_level]])
melts[[taxrank_level]] <- as.character(melts[[taxrank_level]])
# replace all NAs with Other
melts[[taxrank_level]][is.na(melts[[taxrank_level]])] <- "Other"
# convert Phyla back to a factor
melts[[taxrank_level]] <- as.factor(melts[[taxrank_level]])
levels(factor(melts[[taxrank_level]]))
melts$Phylum <- factor(melts$Phylum, levels=sort0)
# sort taxa
summary(melts)
```

Plot relative abundance of the "top 5" Phyla
```{r}
# Build a custom color palette (based on Farrow and Ball paints)
farrowAndBall_swatch <- c(
  "#4d5b6a" #Stiffkey Blue
  ,"#a1c5c8" #Blue Ground
  ,"#686a47" #Bancha
  ,"#ecc363" #Babouche
  ,"#a04344" #Incarnadine  
  ,"#50414c" #Pelt
)

# Specify the stacking order of the top 5 Phyla
sort_order <- c( "p__Proteobacteria",
                "p__Chloroflexi",
                "p__Verrucomicrobia",
                "p__Bacteroidetes",
                "p__Actinobacteria",
                "Other")
df <- melts
df$Phylum <- factor(df$Phylum, levels = rev(sort_order))

# plot relative abundance
plot3 <- ggplot(data=df, aes(x=Group, y=Abundance, fill=factor(Phylum))) + 
  geom_bar(stat="identity",position = "stack") +
  scale_fill_manual(values = rev(farrowAndBall_swatch)) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  guides(fill=guide_legend(title=taxrank_level))
plot3
#ggsave("rel_abun.pdf", plot = plot3, width=6, height=8, useDingbats=FALSE)
#ggsave("rel_abun.png", plot = plot3, width=6, height=8, dpi=600)
```

Actinobacteria subset
```{r}
#Subset Actinobacteria
all.Actino <- subset_taxa(all.trans, Phylum %in% c("p__Actinobacteria"))
all.Actino.ASV.temp <- psmelt(all.Actino)
all.Actino.ASV <- subset(all.Actino.ASV.temp, Abundance > 0)
length(unique(all.Actino.ASV$OTU))

Actino.temp <- as.data.frame(paste(all.Actino.ASV$Group,all.Actino.ASV$Sample_ID,sep="_")) 
Actino.temp$Abundance <- all.Actino.ASV$Abundance
Actino.temp$OTU <- all.Actino.ASV$OTU
colnames(Actino.temp)[1] <- "Group"
Actino.RA <- Actino.temp%>%
  group_by(Group)%>%
  summarise(total=(sum(Abundance)))
Actino.RA
#Remove replicate names from Group
Actino.RA <- data.frame(lapply(Actino.RA, function(x) { gsub("_1", "", x) }))
Actino.RA <- data.frame(lapply(Actino.RA, function(x) { gsub("_2", "", x) }))
Actino.RA <- data.frame(lapply(Actino.RA, function(x) { gsub("_3", "", x) }))
Actino.RA <- data.frame(lapply(Actino.RA, function(x) { gsub("_4", "", x) }))
Actino.RA <- data.frame(lapply(Actino.RA, function(x) { gsub("_5", "", x) }))
Actino.RA[,2] <- as.numeric(as.character(Actino.RA[,2]))
Actino.RA
#write.csv(Actino.RA,"Actino.summary.csv")

#Perform ANOVA and TukeyHSD posthoc
Actino_aov <- aov(total ~ Group, data = Actino.RA)
summary(Actino_aov)
Actino_tukey <- HSD.test(Actino_aov, 'Group')
Actino_tukey

#Plot Actinobacteria relative abundance
p.Actino<- ggplot(data=Actino.RA, aes(x=Group, y=total, fill=Group))+
  geom_boxplot(fill="#a04344")+
  geom_dotplot(binaxis='y',fill="black", stackdir='center', stackratio=1, dotsize=1.25)+
  ylim(0,35)+
  labs(title = "Actinobacteria", x = "Group", y = "Abundance") +
  theme_bw(base_size = 20) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  theme(plot.title = element_text(size = 20, hjust = 0.5))
p.Actino

#ggsave("Actino_sub.pdf", plot = p.Actino, width=5, height=6, useDingbats=FALSE)
#ggsave("Actino_sub.png", plot = p.Actino, width=5, height=6, dpi=600)
```

Bacteroidetes subset
```{r}
#Subset Bacteroidetes
all.Bacteroidetes <- subset_taxa(all.trans, Phylum %in% c("p__Bacteroidetes"))
all.Bacteroidetes.ASV.temp <- psmelt(all.Bacteroidetes)
all.Bacteroidetes.ASV <- subset(all.Bacteroidetes.ASV.temp, Abundance > 0)
length(unique(all.Bacteroidetes.ASV$OTU))

Bacteroidetes.temp <- as.data.frame(paste(all.Bacteroidetes.ASV$Group,all.Bacteroidetes.ASV$Sample_ID,sep="_"))
Bacteroidetes.temp$Abundance <- all.Bacteroidetes.ASV$Abundance
Bacteroidetes.temp$OTU <- all.Bacteroidetes.ASV$OTU
colnames(Bacteroidetes.temp)[1] <- "Group"
Bacteroidetes.RA <- Bacteroidetes.temp%>%
  group_by(Group)%>%
  summarise(total=sum(Abundance))
Bacteroidetes.RA
#Remove replicate names from Group
Bacteroidetes.RA <- data.frame(lapply(Bacteroidetes.RA, function(x) { gsub("_1", "", x) }))
Bacteroidetes.RA <- data.frame(lapply(Bacteroidetes.RA, function(x) { gsub("_2", "", x) }))
Bacteroidetes.RA <- data.frame(lapply(Bacteroidetes.RA, function(x) { gsub("_3", "", x) }))
Bacteroidetes.RA <- data.frame(lapply(Bacteroidetes.RA, function(x) { gsub("_4", "", x) }))
Bacteroidetes.RA <- data.frame(lapply(Bacteroidetes.RA, function(x) { gsub("_5", "", x) }))
Bacteroidetes.RA[,2] <- as.numeric(as.character(Bacteroidetes.RA[,2]))
Bacteroidetes.RA
#write.csv(Bacteroidetes.RA,"Bacteroidetes.summary.csv")

#Perform ANOVA and TukeyHSD posthoc
Bacteroidetes_aov <- aov(total ~ Group, data = Bacteroidetes.RA)
summary(Bacteroidetes_aov)
Bacteroidetes_tukey <- HSD.test(Bacteroidetes_aov, 'Group')
Bacteroidetes_tukey

#Plot Bacteroidetes relative abundance
p.Bacteroidetes<- ggplot(data=Bacteroidetes.RA, aes(x=Group, y=total, fill=Group))+
  geom_boxplot(fill="#ecc363")+
  geom_dotplot(binaxis='y',fill="black", stackdir='center', stackratio=1, dotsize=1.25)+
  ylim(0,40)+
  labs(title = "Bacteroidetes", x = "Group", y = "Abundance") +
  theme_bw(base_size = 20) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  theme(plot.title = element_text(size = 20, hjust = 0.5))
p.Bacteroidetes

#ggsave("Bacteroidetes_sub.pdf", plot = p.Bacteroidetes, width=5, height=6, useDingbats=FALSE)
#ggsave("Bacteroidetes_sub.png", plot = p.Bacteroidetes, width=5, height=6, dpi=600)
```

Verrucomicrobia subset
```{r}
#Subset Verrucomicrobia
all.Verrucomicrobia <- subset_taxa(all.trans, Phylum %in% c("p__Verrucomicrobia"))
all.Verrucomicrobia.ASV.temp <- psmelt(all.Verrucomicrobia)
all.Verrucomicrobia.ASV <- subset(all.Verrucomicrobia.ASV.temp, Abundance > 0)
length(unique(all.Verrucomicrobia.ASV$OTU))

Verrucomicrobia.temp <- as.data.frame(paste(all.Verrucomicrobia.ASV$Group,all.Verrucomicrobia.ASV$Sample_ID,sep="_"))
Verrucomicrobia.temp$Abundance <- all.Verrucomicrobia.ASV$Abundance
Verrucomicrobia.temp$OTU <- all.Verrucomicrobia.ASV$OTU
colnames(Verrucomicrobia.temp)[1] <- "Group"
Verrucomicrobia.RA <- Verrucomicrobia.temp%>%
  group_by(Group)%>%
  summarise(total=sum(Abundance))
Verrucomicrobia.RA
#Remove replicate names from Group
Verrucomicrobia.RA <- data.frame(lapply(Verrucomicrobia.RA, function(x) { gsub("_1", "", x) }))
Verrucomicrobia.RA <- data.frame(lapply(Verrucomicrobia.RA, function(x) { gsub("_2", "", x) }))
Verrucomicrobia.RA <- data.frame(lapply(Verrucomicrobia.RA, function(x) { gsub("_3", "", x) }))
Verrucomicrobia.RA <- data.frame(lapply(Verrucomicrobia.RA, function(x) { gsub("_4", "", x) }))
Verrucomicrobia.RA <- data.frame(lapply(Verrucomicrobia.RA, function(x) { gsub("_5", "", x) }))
Verrucomicrobia.RA[,2] <- as.numeric(as.character(Verrucomicrobia.RA[,2]))
Verrucomicrobia.RA
#write.csv(Verrucomicrobia.RA,"Verrucomicrobia.summary.csv")

#Perform ANOVA and TukeyHSD posthoc
Verrucomicrobia_aov <- aov(total ~ Group, data = Verrucomicrobia.RA)
summary(Verrucomicrobia_aov)
Verrucomicrobia_tukey <- HSD.test(Verrucomicrobia_aov, 'Group')
Verrucomicrobia_tukey

#Plot Verrucomicrobia relative abundance
p.Verrucomicrobia<- ggplot(data=Verrucomicrobia.RA, aes(x=Group, y=total, fill=Group))+
  geom_boxplot(fill="#686a47")+
  geom_dotplot(binaxis='y',fill="black", stackdir='center', stackratio=1, dotsize=1.25)+
  ylim(0,10)+
  labs(title = "Verrucomicrobia", x = "Group", y = "Abundance") +
  theme_bw(base_size = 20) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  theme(plot.title = element_text(size = 20, hjust = 0.5))
p.Verrucomicrobia

#ggsave("Verrucomicrobia_sub.pdf", plot = p.Verrucomicrobia, width=5, height=6, useDingbats=FALSE)
#ggsave("Verrucomicrobia_sub.png", plot = p.Verrucomicrobia, width=5, height=6, dpi=600)
```

Chloroflexi subset
```{r}
#Subset Chloroflexi
all.Chloroflexi <- subset_taxa(all.trans, Phylum %in% c("p__Chloroflexi"))
all.Chloroflexi.ASV.temp <- psmelt(all.Chloroflexi)
all.Chloroflexi.ASV <- subset(all.Chloroflexi.ASV.temp, Abundance > 0)
length(unique(all.Chloroflexi.ASV$OTU))

Chloroflexi.temp <- as.data.frame(paste(all.Chloroflexi.ASV$Group,all.Chloroflexi.ASV$Sample_ID,sep="_"))
Chloroflexi.temp$Abundance <- all.Chloroflexi.ASV$Abundance
Chloroflexi.temp$OTU <- all.Chloroflexi.ASV$OTU
colnames(Chloroflexi.temp)[1] <- "Group"
Chloroflexi.RA <- Chloroflexi.temp%>%
  group_by(Group)%>%
  summarise(total=sum(Abundance))
Chloroflexi.RA
#Remove replicate names from Group
Chloroflexi.RA <- data.frame(lapply(Chloroflexi.RA, function(x) { gsub("_1", "", x) }))
Chloroflexi.RA <- data.frame(lapply(Chloroflexi.RA, function(x) { gsub("_2", "", x) }))
Chloroflexi.RA <- data.frame(lapply(Chloroflexi.RA, function(x) { gsub("_3", "", x) }))
Chloroflexi.RA <- data.frame(lapply(Chloroflexi.RA, function(x) { gsub("_4", "", x) }))
Chloroflexi.RA <- data.frame(lapply(Chloroflexi.RA, function(x) { gsub("_5", "", x) }))
Chloroflexi.RA[,2] <- as.numeric(as.character(Chloroflexi.RA[,2]))
Chloroflexi.RA
#write.csv(Chloroflexi.RA,"Chloroflexi.summary.csv")

#Perform ANOVA and TukeyHSD posthoc
Chloroflexi_aov <- aov(total ~ Group, data = Chloroflexi.RA)
summary(Chloroflexi_aov)
Chloroflexi_tukey <- HSD.test(Chloroflexi_aov, 'Group')
Chloroflexi_tukey

#Plot Chloroflexi relative abundance
p.Chloroflexi<- ggplot(data=Chloroflexi.RA, aes(x=Group, y=total, fill=Group))+
  geom_boxplot(fill="#a1c5c8")+
  geom_dotplot(binaxis='y',fill="black", stackdir='center', stackratio=1, dotsize=1.25)+
  ylim(0,10)+
  labs(title = "Chloroflexi", x = "Group", y = "Abundance") +
  theme_bw(base_size = 20) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  theme(plot.title = element_text(size = 20, hjust = 0.5))
p.Chloroflexi

#ggsave("Chloroflexi_sub.pdf", plot = p.Chloroflexi, width=5, height=6, useDingbats=FALSE)
#ggsave("Chloroflexi_sub.png", plot = p.Chloroflexi, width=5, height=6, dpi=600)
```

Proteobacteria subset
```{r}
#Subset Proteobacteria
all.Proteo <- subset_taxa(all.trans, Phylum %in% c("p__Proteobacteria"))
all.Proteo.ASV.temp <- psmelt(all.Proteo)
all.Proteo.ASV <- subset(all.Proteo.ASV.temp, Abundance > 0)
length(unique(all.Proteo.ASV$OTU))

Proteo.temp <- as.data.frame(paste(all.Proteo.ASV$Group,all.Proteo.ASV$Sample_ID,sep="_"))
Proteo.temp$Abundance <- all.Proteo.ASV$Abundance
Proteo.temp$OTU <- all.Proteo.ASV$OTU
colnames(Proteo.temp)[1] <- "Group"
Proteo.RA <- Proteo.temp%>%
  group_by(Group)%>%
  summarise(total=sum(Abundance))
Proteo.RA
#Remove replicate names from Group
Proteo.RA <- data.frame(lapply(Proteo.RA, function(x) { gsub("_1", "", x) }))
Proteo.RA <- data.frame(lapply(Proteo.RA, function(x) { gsub("_2", "", x) }))
Proteo.RA <- data.frame(lapply(Proteo.RA, function(x) { gsub("_3", "", x) }))
Proteo.RA <- data.frame(lapply(Proteo.RA, function(x) { gsub("_4", "", x) }))
Proteo.RA <- data.frame(lapply(Proteo.RA, function(x) { gsub("_5", "", x) }))
Proteo.RA[,2] <- as.numeric(as.character(Proteo.RA[,2]))
Proteo.RA
#write.csv(Proteo.RA,"Proteo.summary.csv")

#Perform ANOVA and TukeyHSD posthoc
Proteo_aov <- aov(total ~ Group, data = Proteo.RA)
summary(Proteo_aov)
Proteo_tukey <- HSD.test(Proteo_aov, 'Group')
Proteo_tukey

#Plot Proteobacteria relative abundance
p.Proteo<- ggplot(data=Proteo.RA, aes(x=Group, y=total, fill=Group))+
  geom_boxplot(fill="#4d5b6a")+
  geom_dotplot(binaxis='y',fill="black", stackdir='center', stackratio=1, dotsize=1.25)+
  ylim(0,80)+
  labs(title = "Proteobacteria", x = "Group", y = "Abundance") +
  theme_bw(base_size = 20) +
  theme(axis.text.x=element_text(size=12,color="black",angle=90), 
        axis.text.y=element_text(size=12,color="black"), 
        axis.title=element_text(size=12,face="bold"),
        text=element_text(size=12)) +
  theme(plot.title = element_text(size = 20, hjust = 0.5))
p.Proteo

#ggsave("Proteo_sub.pdf", plot = p.Proteo, width=5, height=6, useDingbats=FALSE)
#ggsave("Proteo_sub.png", plot = p.Proteo, width=5, height=6, dpi=600)
```

Here is code that can make a single merged plot with all phyla subsets, using cowplot. 
```{r eval=FALSE, include=TRUE}
(p.Actino| p.Proteo | p.Verrucomicrobia)/(plot3| p.Bacteroidetes | p.Chloroflexi)
# ggsave does not work with cowplot. Export with width=14, height= 12.
```

##Indicator species analysis
The following chunks of code are modified from Simmons et al. 2020
Function #1 needed to run analysis
```{r}
#a function to get indicator species for levels of a given column in your metadata file
#takes a phyloseq object and returns a dataframe of indicators
get_indicators <- function(physeq, groupfactor){
  all.data <- psmelt(physeq)
  all.data <- subset(all.data, Abundance > 0)
  metadata <- sample_data(physeq)
  spec <- subset(all.data, select = c("Sample", "OTU", "Abundance"))
  spec <- reshape2::dcast(spec, Sample ~ OTU, value.var =  "Abundance")
  spec <- mutate_at(spec, c(2:ncol(spec)), ~replace(., is.na(.), 0))
  rownames(spec) <- spec$Sample
  spec <- subset(spec, select = -c(Sample))
  spec <- spec[order(rownames(spec)),]
  clust <- as.data.frame(metadata[[groupfactor]])
  colnames(clust) <- "group"
  clust$group.level <- as.integer(clust$group)
  rownames(clust) <- metadata$Sample_names #change to match metadata
  clust <- subset(clust, rownames(clust) %in% rownames(spec))
  clust <- clust[order(rownames(clust)),]
  if (!(identical(rownames(spec), rownames(clust)))) {
    stop("Species and clustering dataframes are misaligned! Check sample orders.")
  }
  indvals <- indval(spec, clust$group.level, numitr = 10000)
  summary(indvals)
  #convert all.indvals to a single df for ease of access
  pval <- as.data.frame(indvals$pval)
  pval$ASV <- rownames(pval)
  #fdr <- as.data.frame(p.adjust(as.vector(indvals$pval), method = "BH")) #having trouble getting this to work in function. Can do after.
  #fdr$ASV <- rownames(fdr) #change
  maxcls <- as.data.frame(indvals$maxcls)
  maxcls$ASV <- rownames(maxcls)
  indcls <- as.data.frame(indvals$indcls)
  indcls$ASV <- rownames(indcls)
  all.indvals <- merge(pval, maxcls, by = "ASV") #update if using FDR
  all.indvals <- merge(all.indvals, indcls, by = "ASV")
  colnames(all.indvals) <- c("ASV","pval","group.level","indcls") #update if using FDR
  all.indvals <- merge(all.indvals, clust, by = "group.level")
  all.indvals <- subset(all.indvals, select = c(ASV, group, pval, indcls)) #update if using FDR
  all.indvals <- unique(all.indvals)
  #get only significant indicators (p<=0.05) with an indcls >=0.5
  sig.indvals <- subset(all.indvals, pval <= 0.05) #update if using FDR
  sig.indvals <- subset(sig.indvals, indcls >= 0.5)
  #return an object with both the subsetted significant indvals and all indvals
  indvals <- list("all.indvals" = all.indvals, "sig.indvals" = sig.indvals, "all.data" = indvals)
}
```

Function #2 needed to run analysis
```{r}
clean_taxtable <- function(physeq){
  tax.table <- as.data.frame(tax_table(physeq))
  colnames(tax.table) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
  tax.table$Domain <- as.character(lapply(tax.table$Domain, sub, pattern = "k__", replacement = ""))
  tax.table$Phylum <- as.character(lapply(tax.table$Phylum, sub, pattern = "p__", replacement = ""))
  tax.table$Class <- as.character(lapply(tax.table$Class, sub, pattern = "c__", replacement = ""))
  tax.table$Order <- as.character(lapply(tax.table$Order, sub, pattern = "o__", replacement = ""))
  tax.table$Family <- as.character(lapply(tax.table$Family, sub, pattern = "f__", replacement = ""))
  tax.table$Genus <- as.character(lapply(tax.table$Genus, sub, pattern = "g__", replacement = ""))
  tax.table$Species <- as.character(lapply(tax.table$Species, sub, pattern = "s__", replacement = ""))
  tax.table$Domain <- ifelse(tax.table$Domain == "", "Unknown", as.character(tax.table$Domain))
  tax.table$Phylum <- ifelse(tax.table$Phylum == "", "Unknown", as.character(tax.table$Phylum))
  tax.table$Class <- ifelse(tax.table$Class == "", "Unknown", as.character(tax.table$Class))
  tax.table$Order <- ifelse(tax.table$Order == "", "Unknown", as.character(tax.table$Order))
  tax.table$Family <- ifelse(tax.table$Family == "", "Unknown", as.character(tax.table$Family))
  tax.table$Genus <- ifelse(tax.table$Genus == "", "Unknown", as.character(tax.table$Genus))
  tax.table$Species <- ifelse(tax.table$Species == "", "Unknown", as.character(tax.table$Species))
  tax.table$Domain <- ifelse(is.na(tax.table$Domain) == TRUE , "Unknown", as.character(tax.table$Domain))
  tax.table$Phylum <- ifelse(is.na(tax.table$Phylum) == TRUE , "Unknown", as.character(tax.table$Phylum))
  tax.table$Class <- ifelse(is.na(tax.table$Class) == TRUE , "Unknown", as.character(tax.table$Class))
  tax.table$Order <- ifelse(is.na(tax.table$Order) == TRUE , "Unknown", as.character(tax.table$Order))
  tax.table$Family <- ifelse(is.na(tax.table$Family) == TRUE , "Unknown", as.character(tax.table$Family))
  tax.table$Genus <- ifelse(is.na(tax.table$Genus) == TRUE , "Unknown", as.character(tax.table$Genus))
  tax.table$Species <- ifelse(is.na(tax.table$Species) == TRUE , "Unknown", as.character(tax.table$Species))
  tax.table$ASV <- rownames(tax.table)
  rownames(tax.table) <- NULL
  tax.table <- tax.table
}
```

Remove ultralow abundance ASVs to prevent false postives in indicator analysis
```{r}
#we're interested in ASVs that are reliable indicators
#an ASV with 10 (out of 10K!) reads in one or two samples in a category could be a significant indicator simply by virtue of not being found at all in other categories
#removes ASVs with fewer than X total reads and must be present in at least X% of all samples
filter <- phyloseq::genefilter_sample(all.clean, filterfun_sample(function(x) x >= 25), #set at 25
                                      A = 0.05*nsamples(all.clean)-1) #set at 5%
all.clean.filtered <- prune_taxa(filter, all.clean)
TestMelt <- psmelt(all.clean.filtered) #check how many reads are trimmed compared to all.clean.df
rm(TestMelt)
#remove any samples with fewer than 2000 reads, then transform to relative abundance
#all.clean.filtered <- prune_samples(sample_sums(all.clean.filtered)>=2000, all.clean.filtered) #this step is not needed for this study
all.clean.filtered <- transform_sample_counts(all.clean.filtered, function(x) x/sum(x))
tax.table <- clean_taxtable(all.clean.filtered)
```

Run indicator analyses
```{r}
#run indicator analysis for different groupings
TreatmentIndicators <- get_indicators(all.clean.filtered, groupfactor = "Treatment")
GenotypeIndicators <- get_indicators(all.clean.filtered, groupfactor = "Genotype")
GroupIndicators <- get_indicators(all.clean.filtered, groupfactor = "Group")
#subset only  the significant indicators
sig.TreatmentIndicators <- TreatmentIndicators$sig.indvals
sig.GenotypeIndicators <- GenotypeIndicators$sig.indvals
sig.GroupIndicators <- GroupIndicators$sig.indvals
#Add taxonomy information
sig.TreatmentIndicators <- merge(sig.TreatmentIndicators, tax.table, by = "ASV")
sig.GenotypeIndicators <- merge(sig.GenotypeIndicators, tax.table, by = "ASV")
sig.GroupIndicators <- merge(sig.GroupIndicators, tax.table, by = "ASV")
#export csv
#write.csv(sig.TreatmentIndicators, row.names = FALSE, quote = FALSE, file = "sig.TreatmentIndicators.csv")
#write.csv(sig.GenotypeIndicators, row.names = FALSE, quote = FALSE, file = "sig.GenotypeIndicators.csv")
#write.csv(sig.GroupIndicators, row.names = FALSE, quote = FALSE, file = "sig.GroupIndicators.csv")
```





